Source/WebKit/UIProcess/Extensions/WebExtension.cpp

```diff
@@ -1548,6 +1557,279 @@ void WebExtension::populatePermissionsPropertiesIfNeeded()
+        return iconsObject.getString(anyManifestKey);
+#endif
+
+    return iconsObject.getString(String::number(bestSize));
+}
+
+RefPtr<WebCore::Icon> WebExtension::bestImageForIconsDictionaryManifestKey(const JSON::Object& object, const String& manifestKey, WebCore::FloatSize idealSize, IconsCache& cacheLocation, Error error, const String& customLocalizedDescription)
+{
+    auto currentScales = availableScreenScales();
+    if (auto cachedScales = cacheLocation.getOptional("scales"_s)) {
+        auto scales = std::get_if<Vector<double>>(&*cachedScales);
+        if (!scales || *scales != currentScales)
+            cacheLocation.set("scales"_s, IconVariant(currentScales));
+    } else
+        cacheLocation.set("scales"_s, IconVariant(currentScales));
+
+    auto cacheKey = idealSize.toJSONString();
+    if (auto cacheResult = cacheLocation.getOptional(cacheKey)) {
+        if (std::holds_alternative<RefPtr<WebCore::Icon>>(*cacheResult))
+            return *std::get_if<RefPtr<WebCore::Icon>>(&*cacheResult);
+    }
+
+    RefPtr iconObject = object.getObject(manifestKey);
+    if (!iconObject)
+        return nullptr;
+
+    RefPtr result = bestImageInIconsDictionary(iconObject, idealSize, [&](Ref<API::Error> error) {
+        recordError(error);
+    });
+
+    if (!result) {
+        if (iconObject->size()) {
+            // Record an error if the dictionary had values, meaning the likely failure is the images were missing on disk or bad format.
+            recordError(createError(error, customLocalizedDescription));
+        } else if (!iconObject->size() || object.getValue(manifestKey)) {
+            // Record an error if the key had dictionary that was empty, or the key had a value of the wrong type.
+            recordError(createError(error));
+        }
+
+        return nullptr;
+    }
+
+    cacheLocation.set(cacheKey, IconVariant(result));
+
+    return result;
+}
+
+#if ENABLE(WK_WEB_EXTENSIONS_ICON_VARIANTS)
+static OptionSet<WebExtension::ColorScheme> toColorSchemes(RefPtr<JSON::Value> value)
+{
+    using ColorScheme = WebExtension::ColorScheme;
+
+    if (!value) {
+        // A nil or invalid value counts as all color schemes.
+        return { ColorScheme::Light, ColorScheme::Dark };
+    }
+
+    OptionSet<ColorScheme> result;
+
+    RefPtr array = value->asArray();
+    if (!array)
+        return { ColorScheme::Light, ColorScheme::Dark };
+
+    for (Ref value : *array) {
+        if (value->asString() == lightManifestKey)
+            result.add(ColorScheme::Light);
+        if (value->asString() == darkManifestKey)
+            result.add(ColorScheme::Dark);
+    }
+
+    return result;
+}
+
+RefPtr<JSON::Value> WebExtension::iconsDictionaryForBestIconVariant(RefPtr<JSON::Array> variants, size_t idealPixelSize, ColorScheme idealColorScheme)
+{
+    if (!variants || !variants->length())
+        return nullptr;
+
+    if (variants->length() == 1)
+        return variants->begin()->ptr();
```

Comment:
Avoid making an `iterator`.
```suggestion
        return variants->get(0)->ptr();
```